---
title: "Class 9: Structural Bioinformatics (Pt. 1)"
author: "Audrey Nguyen"
format: gfm
---

# Introduction to the RCSB Protein Data Bank (PDB)

## What is in the PDB anyway?

The main database of biomolecular structures is called the PDB and is available at www.rcsb.org.

Let's begin by seeing what is in this database:

## PDB Statistics

Download a CSV file from the PDB site (accessible from "Analyze" > "PDB Statistics" > "by Experimental Method and Molecular Type"). 

> Q1. What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy?

```{r}
pdbstats <- read.csv("Data Export Summary.csv", row.names = 1)
head(pdbstats)
```

```{r}
n.xray <- sum(as.numeric(gsub(",", "", pdbstats$X.ray)))
n.em <- sum(as.numeric(gsub(",", "", pdbstats$EM)))
n.total <- sum(as.numeric(gsub(",", "", pdbstats$Total)))
p.xray <- (n.xray / n.total) * 100
p.em <- (n.em / n.total) * 100
round(p.xray, 2)
round(p.em, 2)
```

There are 172654 (85.9%) protein structures in the X.ray and 14105 (7.02%) protein structures in the Electron Microscopy in the current PDB database. 

> Q2. What proportion of structures in the PDB are protein?

```{r}
as.numeric(gsub(",", "", pdbstats$Total)) / n.total
```
It looks like about 86.9% are protein structures. 

> Q3. Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

It is not straight-forward to find all HIV-1 protease structures using plain text searching on the database. 

# Visualizing the HIV-1 protease structure

> Q4. Water molecules normally have 3 atoms. Why do we see just one atom per molecule in this structure?

Depending on the xray quality, it is hard to see the hydrogen atoms because they're so small.

> Q5. There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have? 

HOH 308 

> Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

![HIV figure](1HSG.png) 

# Introduction to Bio3D in R

We will use the `bio3d` package for this:

```{r}
library(bio3d)
```

## Reading PDB file data into R

```{r}
# accessing online PDB file
pdb <- read.pdb("1hsg")
pdb
```

> Q7. How many amino acid residues are there in this pdb object?

There are 198 amino acid residues. 

> Q8. Name one of the two non-protein residues? 

Water (HOH)

> Q9. How many protein chains are in this structure?

There are 2 protein chains in this structure. 

```{r}
attributes(pdb)
```

```{r}
head(pdb$atom)
```

What is the first residue 3 letter code?

```{r}
pdb$atom$resid[1]  
aa321(pdb$atom$resid[1])
```

## Predicting functional motions of a single structure

```{r}
adk <- read.pdb("6s36")
```

```{r}
adk
```

Normal mode analysis (NMA) is a structural bioinformatics method to predict protein flexibility and potential functional motions (aka conformational changes). 

```{r}
# perform flexibility prediction
m <- nma(adk)
```

```{r}
plot(m)
```

```{r}
mktrj(m, file="adk_m7.pdb")
```

# Comparative structure analysis of Adenylate Kinase

Today we are continuing where we left off last day building towards completing the loop from biomolecular structural data to our new analysis methods like PCA and clustering.

Install bio3d, devtools, and BiocManager (msa).

> Q10. Which of the packages above is found only on BioConductor and not CRAN?

msa is found only on BioConductor.

> Q11. Which of the above packages is not found on BioConductor or CRAN? 

bio3d-view

> Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket? 

True 

## Search and retrieve ADK structures

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
```

```{r}
aa
```

> Q13. How many amino acids are in this sequence?

There are 214 amino acids. 

```{r}
# blast or hmmer search
# b <- blast.pdb(aa)
```

I could save and load my blast results next time so I don't need to run the search every time. 

```{r}
# saveRDS(b, file = "blast_results.RDS")
```

```{r}
b <- readRDS(file = "blast_results.RDS")
```

```{r}
# plot a summary of search results
hits <- plot(b)
```

```{r}
# list out some 'top hits'
head(hits$pdb.id)
```

```{r}
hits <- NULL
hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
```

```{r}
# download related PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

## Align and superpose structures

```{r}
# align related PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile = "msa")
```

```{r}
# vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# draw schematic alignment
# plot(pdbs, labels=ids)
```

Grey regions = aligned residues
White regions = gap regions
Red bar = sequence conservation

## Annotate collected PDB structures

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
```

We can view all available annotation data:

```{r}
anno
```

## Principal component analysis

We will use the `pca()` function from the `bio3d` package as this one is designed to work nicely with biomolecular data.

```{r}
# perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```
These are the results of PCA on Adenylate kinase X-ray structures. Each dot represents one PDB structure. 

We can focus in on PC1 and PC2.

Function `rmsd()` will calculate all pairwise RMSD values of the structural ensemble. This facilitates clustering analysis based on the pairwise structural deviation:

```{r}
# calculate RMSD
rd <- rmsd(pdbs)

# structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k = 3)

plot(pc.xray, 1:2, col = "grey50", bg = grps.rd, pch = 21, cex = 1)
```

# Optional further visualization

```{r}
# visualize first principal component
pc1 <- mktrj(pc.xray, pc = 1, file = "pc_1.pdb")
```

You can view this in Molstar by opening the "pc_1.pdb" file. You can also look at the animations. 

```{r}
# plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1 = pc.xray$z[, 1], PC2 = pc.xray$z[, 2], col = as.factor(grps.rd), ids = ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col = col, label = ids) +
  geom_point(size = 2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

# Normal mode analysis

Function `nma()` provides normal mode analysis (NMA) on both single structures (if given a single PDB input object) or the complete structure ensemble (if provided with a PDBS input object). This facilitates characterizing and comparing flexibility profiles of related protein structures.

```{r}
# NMA of all structures
modes <- nma(pdbs)
```

```{r}
plot(modes, pdbs, col = grps.rd)
```

> Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

The black and colored lines are different at many points. They differ around residues 50 and in between 100 and 150, or basically around where there are higher fluctuations. 




